strawberryfields/engine.py

```diff
@@ -127,870 +78,173 @@
-            # since it uses the == comparison in addition to "is"
-            raise CircuitError("Trying to use a nonexistent measurement result "
-                               "(e.g. before it can be measured).")
-        if self.func is None:
-            return temp[0]
-        return self.func(*temp)
+from .backends.base import (NotApplicableError, BaseBackend)
 
 
 class Engine:
-    r"""Quantum compiler engine.
-
-    Acts as a context manager (and the context itself) for quantum circuits.
-    The contexts may not be nested.
-
-    .. currentmodule:: strawberryfields.engine.Engine
-
-    The quantum circuit is inputted by using the :meth:`~strawberryfields.ops.Operation.__or__`
-    methods of the quantum operations, which call the :meth:`append` method of the Engine.
-    :meth:`append` checks that the register references are valid and then
-    adds a new :class:`.Command` instance to the Engine command queue.
+    r"""Quantum program executor engine.
 
-    :meth:`run` executes the command queue on the chosen backend, and makes
+    :meth:`run` executes :class:`.Program` instances on the chosen backend, and makes
     measurement results available via the :class:`.RegRef` instances.
 
-    The ``New`` and ``Del`` operations modify the quantum register itself by adding
-    and deleting subsystems. The Engine keeps track of these changes as
-    they enter the command queue in order to be able to report register
-    reference errors as soon as they happen.
-
-    The backend, however, only becomes aware of subsystem changes when
-    the circuit is run. See :meth:`reset_queue` and :meth:`reset`.
-
     Args:
-        num_subsystems (int): Number of subsystems in the quantum register.
+        backend (str): name of the backend
+        host    (str): URL of the remote backend host, or None if a local backend is used
     Keyword Args:
         hbar (float): The value of :math:`\hbar` to initialise the engine with, depending on the
             conventions followed. By default, :math:`\hbar=2`. See
             :ref:`conventions` for more details.
     """
     _current_context = None
 
-    def __init__(self, num_subsystems, hbar=2):
-        self.init_num_subsystems = num_subsystems  #: int: initial number of subsystems
-        self.cmd_queue = []       #: list[Command]: command queue
-        #: list[list[Command]]: list of lists of commands that have been run (one list per run)
-        self.cmd_applied = []
-        self.backend = None       #: BaseBackend: backend for executing the quantum circuit
-        #: float: Numerical value of hbar in the (implicit) units of position and momentum.
+    def __init__(self, backend, host=None, *, hbar=2, **kwargs):
+        #: list[Program]: list of Programs that have been run
+        self.run_progs = []
+        #: str: URL of the remote backend host
+        self.host = host
+        #: float: numerical value of hbar in the (implicit) units of position * momentum
```

Comment:
Are these pycharm style type hints? I was chatting to Tom about this today, and the thinking is we will start allowing/highly suggesting Python 3.5 style type hints from now on in our style guide. So these should probably be moved as so:

```python
# list of Programs that have been run
self.run_progs = [] # type: List[Program]
```

Alternatively, it could be written as a attribute docstring which would get picked up by Sphinx automatically:

```python
self.run_progs = [] # type: List[Program]
"""list[Program]: list of Programs that have been run"""
```

Reply:
They are attribute comments which Sphinx does pick up:
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute

Reply:
Oh nice, I didn't know that. I wonder if mypy also picks them up.

Reply:
Ah, it looks like GVM does not support adding doctring type hinting support to mypy 🙁